### PR TITLE
Add bidirectionality to relationship and relationshiptypes

### PR DESCRIPTION
**Pull request by stian-sandvold**
Adds new configuration option for RelationshipType to interpet as bidirectional or not. Add boolean bidirectional to api payload to indicate if a relationship is bidirectional or not.

Issue: DHIS2-5293

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipType.java [hunk 1]
```diff
@@ -52,6 +52,12 @@
 
     private String description;
 
+    private boolean bidirectional = false;
+
+    private String fromToName;
```
Comment (Bekkalizer) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipType.java]: What is the name of a uni-directional relationship?

Comment (Bekkalizer) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipType.java]: answering my own question: fromToName

Comment (stian-sandvold) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/relationship/RelationshipType.java]: fromToName would be the same as "name" is now.

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+ALTER TABLE relationshiptype
```
Comment (ameenhere) [inline dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql]: To maintain consistency, all sql scripts to be in lower case.

File: dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml [hunk 3]
```diff
@@ -21,6 +21,12 @@
 
         <property name="translations" type="jblTranslations"/>
 
+        <property name="bidirectional" column="bidirectional" not-null="true" />
+
+        <property name="fromToName" column="fromToName" not-null="true" />
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml]: Didnt understand this fully. But if the relationshiptype is unidirectional, is the expectation that fromToName will be the same as toFromName and both will not be null ?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/DefaultTrackerAccessManager.java [hunk 4]
```diff
@@ -462,9 +462,9 @@ public DefaultTrackerAccessManager( AclService aclService, TrackerOwnershipManag
         errors.addAll( canWrite( user, from.getProgramInstance(), false ) );
         errors.addAll( canWrite( user, from.getProgramStageInstance(), false ) );
 
-        errors.addAll( canRead( user, to.getTrackedEntityInstance() ) );
-        errors.addAll( canRead( user, to.getProgramInstance(), false ) );
-        errors.addAll( canRead( user, to.getProgramStageInstance(), false ) );
+        errors.addAll( canWrite( user, to.getTrackedEntityInstance() ) );
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/DefaultTrackerAccessManager.java]: This should be a separate PR 😝 jk

File: dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml (hunk 3 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml]: No, only from-to name is required, since there is always atleast one name/direction. toFrom is the reverse and is optional and will be null for non-bidirectional relationships

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/DefaultTrackerAccessManager.java (hunk 4 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/DefaultTrackerAccessManager.java]: :P

File: dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml (hunk 3 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml]: So the property below should be (not-null="false") ?

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql [hunk 5]
```diff
@@ -0,0 +1,15 @@
+alter table relationshiptype
+  add column bidirectional boolean;
```
Comment (ameenhere) [inline dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql]: Would be good to make the entire script idempotent by adding "add column if not exists"

File: dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml (hunk 3 above)
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/relationship/hibernate/RelationshipType.hbm.xml]: yes :)

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql (hunk 5 above)
Comment (stian-sandvold) [reply dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.32/V2_32_14__Update_relationshiptype_bidirectional.sql]: Will do
